### PR TITLE
Added intake breakbeam

### PR DESCRIPTION
**Pull request by DensityInfinite**
Added the intake component, based on Amber's previous code.
The `do_intake` and `end_intake` methods controls the intake.
There is support for a break beam sensor.

### CODE REVIEW

**Review by Pokesi — CHANGES_REQUESTED** [review]
this should be a controller that uses the component and not a component

File: components/intake.py [hunk 1]
```diff
@@ -1,14 +1,15 @@
...
 import ids
 
 
 class Intake:
     intake_speed = tunable(0.5)
 
-    def __init__(self):
+    def __init__(self) -> None:
         self.deployed = False
+        self.break_beam = DigitalInput(0)
```
Comment (OliverW10) [inline components/intake.py:12]: could you add a DioChannels class in ids.py and add this to it

**Comment by DensityInfinite** [comment]
> this should be a controller that uses the component and not a component

This is now addressed as we are making this a component, removing all the controller features.

File: components/intake.py [hunk 2]
```diff
@@ -18,16 +19,19 @@ def __init__(self):
...
+
+    def end_intake(self) -> None:
+        self.deployed = False
+
+    def execute(self) -> None:
         if self.deployed:
             self.piston.set(DoubleSolenoid.Value.kForward)
             self.motor.set(self.intake_speed)
-        else:
+        elif not self.deployed:
```
Comment (OliverW10) [inline components/intake.py:32]: just else would be fine

File: robot.py [hunk 3]
```diff
@@ -37,12 +37,7 @@ def teleopPeriodic(self) -> None:
         drive_y = -rescale_js(self.gamepad.getLeftX(), 0.1) * Chassis.max_wheel_speed
         drive_z = -rescale_js(self.gamepad.getRightX(), 0.1, exponential=2) * spin_rate
         local_driving = self.gamepad.getBButton()
-
-        if self.gamepad.getYButtonPressed():
```
Comment (OliverW10) [inline robot.py:41]: could you restore these, also in the past left/right bumpers have been handy for deploying/retracting intake

**Review by OliverW10 — COMMENTED** [review]
Looks pretty good, just a few comments

**Review by auscompgeek — COMMENTED** [review]
please re-request review after resolving merge conflicts and addressing Oliver's comments

File: robot.py [hunk 4]
```diff
@@ -21,7 +20,6 @@ class MyRobot(magicbot.MagicRobot):
     vision: Vision
     arm: Arm
     intake: Intake
-    gripper: Gripper
```
Comment (auscompgeek) [inline robot.py:24]: ???

File: components/intake.py [hunk 5]
```diff
@@ -18,16 +20,19 @@ def __init__(self):
             ids.PcmChannels.Intake.intake_piston_reverse,
         )
 
-    def execute(self):
+    def do_intake(self) -> None:
+        self.deployed = True
+
+    def end_intake(self) -> None:
```
Comment (auscompgeek) [inline components/intake.py:26]: These are _very_ unusual names. I don't think it makes sense to tell an intake subsystem to "do intake" or "end intake".

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: components/intake.py (hunk 5 above)
Comment (DensityInfinite) [reply components/intake.py:26]: Should I change it to "intake" and "retract"?

Comment (auscompgeek) [reply components/intake.py:26]: Is there a reason they were renamed from "deploy"/"retract", especially since the variable is still called "deployed"?

Comment (DensityInfinite) [reply components/intake.py:26]: I renamed it when writing a controller for the intake at the very start.
Now that controller doesn't exist anymore I will change it back to "deploy" and "retract", if that is better?

File: robot.py (hunk 4 above)
Comment (DensityInfinite) [reply robot.py:24]: Yeah that was a big mistake. I think I accidentally deleted the import line and this.
Will address this in a moment.

**Comment by DensityInfinite** [comment]
> Looks pretty good, just a few comments. Also could you have it set deployed to false when the break beam is broken

Should I put this in the scoring controller or just in the component?

File: components/intake.py [hunk 6]
```diff
@@ -18,16 +20,19 @@ def __init__(self):
...
             self.piston.set(DoubleSolenoid.Value.kReverse)
             self.motor.set(0.0)
 
-    def deploy(self):
-        self.deployed = True
-
-    def retract(self):
-        self.deployed = False
+    def get_break_beam_sensor_state(self) -> bool:
+        return self.break_beam.get()
```
Comment (auscompgeek) [inline components/intake.py:38]: Please test the break beam sensor on a real control system before merging this.

The roboRIO has an internal pull-up resistor on the DIO pins, so the switch being tripped will typically (i.e. unless the switch being unplugged should behave the same way as tripping the switch) connect signal to ground, pulling the pin low.

File: robot.py [hunk 7]
```diff
@@ -42,47 +42,24 @@ def teleopPeriodic(self) -> None:
         drive_y = -rescale_js(self.gamepad.getLeftX(), 0.1) * Chassis.max_wheel_speed
         drive_z = -rescale_js(self.gamepad.getRightX(), 0.1, exponential=2) * spin_rate
         local_driving = self.gamepad.getBButton()
+        self.gamepad.getXButton()
 
-        if self.gamepad.getLeftBumper():
+        if self.gamepad.getYButtonPressed():
+            self.intake.deploy()
```
Comment (auscompgeek) [inline robot.py:48]: Why are there multiple paths to deploy?

File: robot.py [hunk 8]
```diff
@@ -42,47 +42,24 @@ def teleopPeriodic(self) -> None:
...
+        if self.gamepad.getLeftBumper():
             self.intake.deploy()
 
-        if self.gripper.game_piece_in_reach():
-            self.gripper.close()
-
-        if self.gamepad.getXButton():
-            self.gripper.open()
+        if self.gamepad.getRightBumper():
+            self.intake.retract()
```
Comment (auscompgeek) [inline robot.py:57]: Why are there multiple paths to retract?

File: components/intake.py (hunk 6 above)
Comment (DensityInfinite) [reply components/intake.py]: Ok, it now reverses on return.

**Comment by OliverW10** [comment]
> Should I put this in the scoring controller or just in the component?

In the component I think

**Comment by OliverW10** [comment]
> Should I put this in the scoring controller or just in the component?

In the component I think

**Review by auscompgeek — APPROVED** [review]
Intake component and `ids` LGTM

File: components/intake.py [hunk 9]
```diff
@@ -1,14 +1,16 @@
 from rev import CANSparkMax
-from wpilib import DoubleSolenoid, PneumaticsModuleType
+from wpilib import DoubleSolenoid, DigitalInput, PneumaticsModuleType
 from magicbot import tunable
 import ids
 
 
 class Intake:
+    ids = ids
```
Comment (auscompgeek) [inline components/intake.py:8]: Wait, what's this doing here?

Comment (DensityInfinite) [reply components/intake.py:8]: There were multiple classes in ids.py so I thought I could access them and their member variables by doing this.
How should have I done it?

Comment (Pokesi) [reply components/intake.py:8]: lmao

File: components/intake.py [hunk 10]
```diff
@@ -1,14 +1,16 @@
...
     intake_speed = tunable(0.5)
 
-    def __init__(self):
+    def __init__(self) -> None:
         self.deployed = False
+        self.break_beam = DigitalInput(self.ids.DioChannels.Intake.break_beam_sensor)
```
Comment (Pokesi) [inline components/intake.py]: use `ids.` not `self.ids.`

File: components/intake.py (hunk 9 above)
Comment (DensityInfinite) [reply components/intake.py]: Addressed in commit. (hahah)

**Review by Pokesi — APPROVED** [review]
`ids = ids` lmao

File: components/intake.py (hunk 10 above)
Comment (DensityInfinite) [reply components/intake.py]: Addressed in the previous commit XD

**Comment by DensityInfinite** [comment]
> `ids = ids` lmao

best thing happened in this branch

**Review by OliverW10 — APPROVED** [review]

**Pull request merged**